extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -356,6 +356,38 @@ export class Bind {
     return this.setStatePromise_;
   }
 
+  /**
+   * Merges a state object into the current global state, or parses a state
+   * expression and merges it into the global state.
+   * @param {(!JsonObject|string)} state
+   * @return {!Promise}
+   */
+  setStateAndUpdateHistory(state) {
+    // Support expressions
+    if (typeof state === 'string') {
+      // Emulate UIEvent 'click'
+      return this.pushStateWithExpression(
+        state,
+        /** @type {!JsonObject} */ ({event: 1})
+      );
+    }
+    // Sanitize and copy state
+    const stateCopy = this.copyJsonObject_(state);
+    if (!stateCopy) {
+      return Promise.resolve();
+    }
+    dev().info(TAG, 'setState:', stateCopy);
+    this.setStatePromise_ = this.setState(stateCopy)
+      .then(() => this.getDataForHistory_())
+      .then(data => {
+        // Don't bother calling History.replace with empty data.
+        if (data) {
+          this.history_.replace(data);
+        }
+      });
+    return this.setStatePromise_;
```

Comment:
Would be nice to de-dupe this code snippet with `setStateWithExpression`. Feel free to create a new private helper method or just drop a `TODO(choumx)` here for me.